vault-plugins/shared/memdb/memdb.go

```diff
@@ -0,0 +1,407 @@
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		default:
+			return t.checkSingleValueOfForeignKey(checkedFieldValue, key, allowedArchivingTimeStamp, allowedArchivingHash)
+		}
+	}
+}
+
+func (t *Txn) checkSingleValueOfForeignKey(singleCheckedFieldValue interface{}, key Relation,
+	allowedArchivingTimeStamp UnixTime, allowedArchivingHash int64) error {
+	relatedRecord, err := t.First(key.RelatedDataType, key.RelatedDataTypeFieldIndexName, singleCheckedFieldValue)
+	if err != nil {
+		return fmt.Errorf("getting related record:%w", err)
+	}
+	if relatedRecord == nil {
+		return fmt.Errorf("FK violation: %q not found at table %q at index %q",
+			singleCheckedFieldValue, key.RelatedDataType, key.RelatedDataTypeFieldIndexName)
+	}
+	a, ok := relatedRecord.(Archivable)
+	if !ok {
+		if allowedArchivingTimeStamp == 0 && allowedArchivingHash == 0 {
+			return nil // related record is not archivable, exists, an no need to check
+		} else {
+			return fmt.Errorf("%w related record %#v is not archivable", ErrNotArchivable, relatedRecord)
+		}
+	}
+	s, h := a.ArchiveMarks()
+	if a.Archived() && (s != allowedArchivingTimeStamp || h != allowedArchivingHash) {
+		return fmt.Errorf("FK violation: %q not found at table %q at index %q",
+			singleCheckedFieldValue, key.RelatedDataType, key.RelatedDataTypeFieldIndexName)
+	}
+	return nil
+}
+
+func (t *Txn) checkCascadeDeletesAndCheckingRelations(table string, obj interface{}) error {
+	rels := append(t.schema.CascadeDeletes[table], t.schema.CheckingRelations[table]...) //nolint:gocritic
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+func (t *Txn) checkCheckingRelations(table string, obj interface{}) error {
+	rels := t.schema.CheckingRelations[table]
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+// implement main loop checking relations
+func (t *Txn) processRelations(relations []Relation, obj interface{},
+	relationHandler func(originObjectFieldValue interface{}, key Relation) error, relationHandlerError error) error {
+	valueIface := reflect.ValueOf(obj)
+	if valueIface.Type().Kind() != reflect.Ptr {
+		return fmt.Errorf("obj `%s` is not ptr", valueIface.Type())
+	}
+	msgs := []string{}
+	for _, key := range relations {
+		field := valueIface.Elem().FieldByName(key.OriginalDataTypeFieldName)
+		if !field.IsValid() {
+			return fmt.Errorf("obj `%s` does not have the field `%s`", valueIface.Type(), key.OriginalDataTypeFieldName)
+		}
+		checkedFieldValue := field.Interface()
+		if err := relationHandler(checkedFieldValue, key); err != nil {
+			msgs = append(msgs, err.Error())
+		}
+	}
+	builder := strings.Builder{}
+	if len(msgs) != 0 {
+		for _, msg := range msgs {
+			if builder.Len() != 0 {
+				builder.WriteString(";")
+			}
+			builder.WriteString(msg)
+		}
+		return fmt.Errorf(builder.String()+":%w", relationHandlerError)
+	}
```

Comment:
I think we should have single common error collecting type for this. Formatting error messages is not a particular feature of a relation processing. (an example of the idea https://play.golang.org/p/eztQmlaleB from [SO](https://stackoverflow.com/questions/33470649/combine-multiple-error-strings))

Reply:
fixed